lib/models/watch-detector.js

```diff
@@ -0,0 +1,238 @@
+var WATCHMAN_INFO = 'Visit http://ember-cli.com/user-guide/#watchman for more info.';
+var semver = require('semver');
+var path = require('path');
+
+var POLLING = 'polling';
+var WATCHMAN = 'watchman';
+var NODE = 'node';
+
+var debug   = require('debug')('ember-cli:watcher');
+
+function WatchPreference(watcher) {
+  this.watcher = watcher || null;
+  this._watchmanInfo = {
+    enabled: false,
+    version: null,
+    canNestRoots: false
+  };
+}
+
+WatchPreference.prototype.watchmanWorks = function(details) {
+  this._watchmanInfo.enabled = true;
+  this._watchmanInfo.version = details.version;
+  this._watchmanInfo.canNestRoots = semver.satisfies(details.version, '>= 3.7.0');
+};
+
+Object.defineProperty(WatchPreference.prototype, 'watchmanInfo', {
+  get: function() {
+    return this._watchmanInfo;
+  }
+});
+
+module.exports = WatchDetector;
+/*
+ * @public
+ *
+ * A testable class that encapsulates which watch technique to use.
+ *
+ */
+function WatchDetector(options) {
+  this.childProcess = options.childProcess;
+  this.watchmanSupportsPlatform = options.watchmanSupportsPlatform;
+  this.fs = options.fs;
+
+  this.ui = options.ui;
+  this.tmp = undefined;
+
+  // This exists, because during tests repeadelty testing the same directories
+  this._doesNodeWorkCache = options.cache || {};
+}
+
+/*
+ * @private
+ *
+ * implements input options parsing and fallback.
+ *
+ * @method extractPreferenceFromOptions
+ * @returns {WatchDetector)
+*/
+WatchDetector.prototype.extractPreferenceFromOptions = function(options) {
+  if (options.watcher === POLLING) {
+    debug('skip detecting watchman, "polling" was selected.');
+    return new WatchPreference(POLLING);
+  } else if (options.watcher === NODE) {
+    debug('skip detecting watchman, "node" was selected.');
+    return new WatchPreference(NODE);
+  } else {
+    debug('no watcher preference set, lets attempt watchman');
+    return new WatchPreference(WATCHMAN);
+  }
+}
+
+/*
+ * @public
+ *
+ * Detect if `fs.watch` provided by node is capable of observing a directory
+ * within `process.cwd()`. Although `fs.watch` is provided as a part of the
+ * node stdlib, their exists platforms it ships to that do not implemented the
+ * required sub-system. For example, there exists `posix` targets, without both
+ * inotify and FSEvents
+ *
+ * @method testIfNodeWatcherAppearsToWork
+ * @returns {Boolean) outcome
+*/
+WatchDetector.prototype.testIfNodeWatcherAppearsToWork  = function() {
+  return new Promise(function(resolve) {
+    var key = path.dirname(path.dirname(process.cwd()));
+
+    if (key in this._doesNodeWorkCache) {
+      resolve(this._doesNodeWorkCache[key]);
+      return;
+    }
+
+    this._doesNodeWorkCache[key] = false;
+    try {
+      // builds a tmp directory at process.cwd() + '/tmp/' + 'something-unique';
+      var tmpDir = quickTemp.makeOrRemake(this, 'tmp');
+
+      this.fs.watch(tmpDir, { persistent: false, recursive: false });
+      this.fs.unwatch(tmpDir);
```

Comment:
some future PR could, this one merely makes sure the call does throw. Like it does on platforms that don't have support.